src/sourmash/index.py

```diff
@@ -219,7 +220,66 @@ def select(self, **kwargs):
             if select_signature(ss, **kwargs):
                 siglist.append(ss)
 
-        return LinearIndex(siglist, self.filename)
+        return LinearIndex(siglist, self.location)
+
+
+class ZipFileLinearIndex(Index):
+    """\
+    A read-only collection of signatures in a zip file.
+
+    Does not support `insert` or `save`.
+    """
+    is_database = True
+
+    def __init__(self, zf, selection_dict=None,
+                 traverse_yield_all=False):
+        self.zf = zf
+        self.selection_dict = selection_dict
+        self.traverse_yield_all = traverse_yield_all
+
+    def __len__(self):
+        return len(list(self.signatures()))
+
+    @property
+    def location(self):
+        return self.zf.filename
+
+    def insert(self, signature):
+        raise NotImplementedError
+
+    def save(self, path):
+        raise NotImplementedError
+
+    @classmethod
+    def load(cls, location, traverse_yield_all=False):
+        "Class method to load a zipfile."
+        zf = zipfile.ZipFile(location, 'r')
+        return cls(zf, traverse_yield_all=traverse_yield_all)
+
+    def signatures(self):
+        "Load all signatures in the zip file."
+        from .signature import load_signatures
+        for zipinfo in self.zf.infolist():
+            # should we load this file? if it ends in .sig OR we are forcing:
+            if zipinfo.filename.endswith('.sig') or \
+               zipinfo.filename.endswith('.sig.gz') or \
+               self.traverse_yield_all:
+                fp = self.zf.open(zipinfo)
+
+                # now load all the signatures and select on ksize/moltype:
+                selection_dict = self.selection_dict
+                for ss in load_signatures(fp):
+                    if selection_dict:
+                        if select_signature(ss, **self.selection_dict):
+                            yield ss
+                    else:
+                        yield ss
```

Comment:
trying to wrap my head around `traverse_yield_all` --

do all files get loaded into generator, but non-sig files are never returned bc they can't be loaded as sigs (and thus won't be selected)?

Reply:
correct!

Reply:
although, looking at it, I'm a little surprised this works. Will dig.

Reply:
ok, commented in 56184ffc - `load_signatures` will fail silently on invalid JSON/files that aren't loadable, and return nothing. (If you pass in `do_raise=True` it will raise a `ValueError`)